src/dmd/errors.d

```diff
@@ -648,17 +648,32 @@ private void _vdeprecationSupplemental(const ref Loc loc, const(char)* format, v
         verrorPrint(loc, Classification.deprecation, "       ", format, ap);
 }
 
+/**
+ * Thrown on a fatal error.
+ */
+class FatalError : Error
+{
+    @safe pure nothrow this(string msg, string file = __FILE__, size_t line = __LINE__)
+    {
+        super(msg, file, line, null);
+    }
+}
+
 /**
  * Call this after printing out fatal error messages to clean up and exit
  * the compiler.
+ * Note: On DMD as a library this won't halt the program, and instead throw
+ * FatalError.
  */
 extern (C++) void fatal()
 {
-    version (none)
+    version (DMDLIB)
+        throw new FatalError("Fatal error", __FILE__, __LINE__);
```

Comment:
should this really be throwing an `Error`? IMO this is no improvement over exit, because what DMD deem fatal, e.g. too many errors, is not irrecoverable for an application using DMD. 

Ideally for an application the process would be catch and exception reset the compiler state and carry on with whatever it was doing.

Reply:
Do you have a better alternative? Throwing exceptions would require breaking a lot of function signatures just for `libdmd`. Propagating the error would require a lot of changes, and possibly make normal flow slow.

I'm still afraid of the possible side effects of throwing a `FatalError` although my brain can't find a clear example of how dangerous this could be, from the available options this seems to be the best one.

As pointed out by @rainers, `onAssertErrorMsg` throws `AssertError` and this can be confused by an `assert`, which the developer may not want to catch, hence the new `FatalError`. One advantage of `onAssertErrorMsg` is that it allocates on a shared TLS space and, in recent versions, it doesn't compute the class instance size at compile time. I can do a change that mimics that behaviour but will require more TLS space (which I guess it's fine).